pennylane/devices/qubit/measure.py

```diff
@@ -86,17 +88,21 @@ def csr_dot_products(
         TensorLike: the result of the measurement
     """
     total_wires = len(state.shape) - is_state_batched
-    Hmat = measurementprocess.obs.sparse_matrix(wire_order=list(range(total_wires)))
 
     if is_state_batched:
+        Hmat = measurementprocess.obs.sparse_matrix(wire_order=list(range(total_wires)))
         state = math.toarray(state).reshape(math.shape(state)[0], -1)
 
         bra = csr_matrix(math.conj(state))
         ket = csr_matrix(state)
         new_bra = bra.dot(Hmat)
         res = new_bra.multiply(ket).sum(axis=1).getA()
-
+    elif is_pauli_sentence(measurementprocess.obs):
+        state = math.toarray(state).flatten()
+        ps = pauli_sentence(measurementprocess.obs)
```

Comment:
If the conversion from hamiltonian to pauli sentence is being a bottleneck, I'd think it might be time to start caching the pauli sentence in the `_pauli_rep` property on Hamiltonians.

Reply:
Could do. For the tests cases I was interested in, changing the `reduce` statement to in-place add like you suggested pretty much solved it.